### PR TITLE
String import/export specifier

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Implements https://github.com/tc39/ecma262/pull/2154
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2396
| Any Dependency Changes?  | `@babel/helper-module-transforms` now depends on `@babel/helper-validator-identifier`
| License                  | MIT

This PR implements support of Arbitrary module namespace identifier names behind `moduleStringNames` parser plugin
```js
import { "some imports" as foo } from "module"
export { foo as "some exports" } from "module"
export { "another imports" as "another exports" } from "module"
export { "rest exports" } from "module"
```
Babel can now parse such code and transform it to CommonJS/AMD/UMD/SystemJS code. Take CommonJS as an output example

```js
const m = require("module");
foo = m["some imports"];
exports["some exports"] = m.foo;
exports["another exports"] = m["another imports"];
exports["rest exports"] = m["rest exports"];
```

A new package `@babel/plugin-syntax-module-string-names` is introduced. See also https://github.com/babel/babel/pull/12091/files#diff-ef433cb06c064c1e7a5af0f448d98a78 for AST spec changes.

Todo:
- [x] umd-transform
- [x] amd-transform
- [x] systemjs-transform (/cc @guybedford)

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12091"></a>

### CODE REVIEW

File: packages/babel-plugin-transform-modules-systemjs/test/fixtures/systemjs/export-named-string/output.mjs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+System.register([], function (_export, _context) {
+  "use strict";
+
+  var foo, bar;
+
+  _export({
+    foo: void 0,
+    bar: void 0
+  });
```
Comment (JLHwung) [inline packages/babel-plugin-transform-modules-systemjs/test/fixtures/systemjs/export-named-string/output.mjs:9]: This output is incorrect, it should be
```js
_export({
    "default exports": foo,
    bar: void 0
  });
```
However it is a regression introduced in https://github.com/babel/babel/pull/9589/files#diff-a0b4df402be122c3213c96111b909761. I will address it in a separate PR.

Comment (JLHwung) [reply packages/babel-plugin-transform-modules-systemjs/test/fixtures/systemjs/export-named-string/output.mjs:9]: Resovled in #12110.

**Review by guybedford — APPROVED** [review]
SystemJS output looks great, and glad it fixed some bugs too.

File: packages/babel-parser/src/parser/statement.js [hunk 2]
```diff
@@ -2216,17 +2242,20 @@ export default class StatementParser extends ExpressionParser {
...
-      this.checkReservedWord(
-        specifier.imported.name,
-        specifier.start,
-        true,
-        true,
-      );
-      specifier.local = specifier.imported.__clone();
+      const { imported } = specifier;
+      if (imported.type === "StringLiteral") {
+        throw this.raise(
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/statement.js:2251]: Why do we need to throw here?

**Review by nicolo-ribaudo — COMMENTED** [review]
I think that we should introduce this behind a syntax plugin. Even if the spec PR has  consensus, it will still take a while to be merged because it doesn't have tests.

Also, for example https://github.com/tc39/ecma262/pull/1174 still took a few months after having consensus and tests.

File: packages/babel-parser/src/parser/statement.js (hunk 2 above)
Comment (JLHwung) [reply packages/babel-parser/src/parser/statement.js:2251]: We don't see an `as` here, so we are sure it is parsed as `import { "foo" }`, which is invalid.

**Review by nicolo-ribaudo — APPROVED** [review]
Except for a question (maybe small error) this looks good :+1:

File: packages/babel-parser/ast/spec.md [hunk 3]
```diff
@@ -1352,21 +1352,24 @@ interface ExportNamedDeclaration <: ModuleDeclaration {
...
 
 ### ExportSpecifier
 
 ```js
 interface ExportSpecifier <: ModuleSpecifier {
   type: "ExportSpecifier";
-  exported: Identifier;
-  local?: Identifier;
+  exported: Identifier | StringLiteral;
+  local?: Identifier | StringLiteral;
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-parser/ast/spec.md:1372]:
```js
export { "😄" as "😂" } from "emojis"
```

File: packages/babel-parser/src/parser/statement.js (hunk 2 above)
Comment (JLHwung) [reply packages/babel-parser/src/parser/statement.js:2261]: I will add a testcase for this error.

File: packages/babel-parser/src/parser/error-message.js [hunk 4]
```diff
@@ -43,6 +43,8 @@ export const ErrorMessages = Object.freeze({
   DuplicateRegExpFlags: "Duplicate regular expression flag",
   ElementAfterRest: "Rest element must be last element",
   EscapedCharNotAnIdentifier: "Invalid Unicode escape",
+  ExportBindingIsString:
+    "Can not use string literal as exported binding without `from`.\n- Did you mean `export { %0 as '%1' } from 'some-module'`",
```
Comment (kaicataldo) [inline packages/babel-parser/src/parser/error-message.js:47]: Looks like we're missing a question mark here:

```suggestion
    "Can not use string literal as exported binding without `from`.\n- Did you mean `export { %0 as '%1' } from 'some-module'`?",
```

**Review by kaicataldo — APPROVED** [review]

File: packages/babel-parser/src/parser/error-message.js [hunk 5]
```diff
@@ -43,6 +43,8 @@ export const ErrorMessages = Object.freeze({
   DuplicateRegExpFlags: "Duplicate regular expression flag",
   ElementAfterRest: "Rest element must be last element",
   EscapedCharNotAnIdentifier: "Invalid Unicode escape",
+  ExportBindingIsString:
+    "Can not use string literal as exported binding without `from`.\n- Did you mean `export { %0 as '%1' } from 'some-module'`?",
```
Comment (existentialism) [inline packages/babel-parser/src/parser/error-message.js:47]: Nit:

```suggestion
    "Cannot use string literal as exported binding without `from`.\n- Did you mean `export { %0 as '%1' } from 'some-module'`?",
```

File: packages/babel-parser/src/parser/error-message.js [hunk 6]
```diff
@@ -53,6 +55,8 @@ export const ErrorMessages = Object.freeze({
   IllegalLanguageModeDirective:
     "Illegal 'use strict' directive in function with non-simple parameter list",
   IllegalReturn: "'return' outside of function",
+  ImportBindingIsString:
+    'Can not use string literal as an imported binding.\n- Did you mean `import { "%0" as foo }`?',
```
Comment (existentialism) [inline packages/babel-parser/src/parser/error-message.js:59]: Nit:

```suggestion
    'A string literal cannot be used as an imported binding.\n- Did you mean `import { "%0" as foo }`?',
```

File: packages/babel-parser/src/parser/error-message.js [hunk 7]
```diff
@@ -95,6 +99,8 @@ export const ErrorMessages = Object.freeze({
     "Only string literals are allowed as module attribute values",
   ModuleAttributesWithDuplicateKeys:
     'Duplicate key "%0" is not allowed in module attributes',
+  ModuleExportNameHasLoneSurrogate:
+    "Export name can not include lone surrogate, but '\\u%0' is seen",
```
Comment (existentialism) [inline packages/babel-parser/src/parser/error-message.js:103]: Maybe something like?

```suggestion
    "An export name cannot be a lone surrogate, found '\\u%0'",
```

Comment (JLHwung) [reply packages/babel-parser/src/parser/error-message.js:103]: A lone surrogate refers to a character falling into non-paired surrogates. An export name is a string thus conceptually an export name can _include_ some character.

**Pull request merged**